src/main/java/org/reaktivity/nukleus/amqp/internal/stream/AmqpServerFactory.java

```diff
@@ -0,0 +1,512 @@
+            DirectBuffer header,
+            int offset,
+            int length)
+        {
+            final DataFW data = dataRW.wrap(writeBuffer, 0, writeBuffer.capacity())
+                .routeId(routeId)
+                .streamId(replyId)
+                .trace(supplyTraceId.getAsLong())
+                .groupId(0)
+                .padding(replyPadding)
+                .payload(header, offset, length)
+                .build();
+            receiver.accept(data.typeId(), data.buffer(), data.offset(), data.sizeof());
+        }
+
+        private void onNetwork(
+            int msgTypeId,
+            DirectBuffer buffer,
+            int index,
+            int length)
+        {
+            switch (msgTypeId)
+            {
+                case BeginFW.TYPE_ID:
+                    final BeginFW begin = beginRO.wrap(buffer, index, index + length);
+                    onBegin(begin);
+                    break;
+                case DataFW.TYPE_ID:
+                    final DataFW data = dataRO.wrap(buffer, index, index + length);
+                    onData(data);
+                    break;
+                case EndFW.TYPE_ID:
+                    final EndFW end = endRO.wrap(buffer, index, index + length);
+                    onEnd(end);
+                    break;
+                case AbortFW.TYPE_ID:
+                    final AbortFW abort = abortRO.wrap(buffer, index, index + length);
+                    onAbort(abort);
+                    break;
+                case WindowFW.TYPE_ID:
+                    final WindowFW window = windowRO.wrap(buffer, index, index + length);
+                    onWindow(window);
+                    break;
+                case ResetFW.TYPE_ID:
+                    final ResetFW reset = resetRO.wrap(buffer, index, index + length);
+                    onReset(reset);
+                    break;
+                case SignalFW.TYPE_ID:
+                    final SignalFW signal = signalRO.wrap(buffer, index, index + length);
+                    onSignal(signal);
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        private void onBegin(
+            BeginFW begin)
+        {
+            doBegin(supplyTraceId.getAsLong());
+        }
+
+        private void onData(
+            DataFW data)
+        {
+            initialBudget -= data.length() + data.padding();
+
+            if (initialBudget < 0)
+            {
+                doReset(supplyTraceId.getAsLong());
+            }
+            else
+            {
+                decodeTraceId = data.trace();
+                final OctetsFW payload = data.payload();
+                final DirectBuffer buffer = payload.buffer();
+
+                int offset = payload.offset();
+                int length = payload.sizeof();
+
+                while (length > 0)
+                {
+                    int consumed = decodeState.decode(buffer, offset, length);
```

Comment:
We also need to detect no change in `decodeState` and break the loop, even if length is still non-zero.
This represents the case where we have insufficient data to completely decode a frame.
Any unprocessed payload then needs to be stored in a buffer slot for the stream, and subsequent `DATA` frames need to have their payload appended at the end of any data in the buffer slot, so that the current decode state can try again when more payload arrives from the network.